### ISSUE TITLE
introspecting Docopt usage strings

### ISSUE THREAD

**Issue by BurntSushi** [docopt/docopt.rs#88]
It seems like there is an opportunity to expose some useful information about Docopt usage strings: http://stackoverflow.com/questions/28210260/parsing-docopt-argument-types-any-language/28214541?noredirect=1#comment44799324_28214541

The parser in this implementation currently drops it on the floor.

**Comment by con-f-use** [comment]
Doesn't this type of introspection raise problems? E.g. if you have somthing like:

```
Usage: myprog --coordinates ( (<cartesian_x> <cartesian_y>) | (<polar_angle> <polar_radius>) )
```

There is no way for docopt to determine wheter in the call `myprog --coordinates 30 72` the coordinates are of type "cartesian" or "polar" faux-type.

**Comment by BurntSushi** [comment]
Those are positional arguments. The SO post is asking for the name of a flag argument. e.g., for `--flag=FILE` extract `FILE` for flag `flag`.

**Comment by con-f-use** [comment]
But `--flag=FILE` is equivalent to `--flag <file>`.

**Comment by BurntSushi** [comment]
Yes... That should work too. Your example contains a flag with *no arguments* and two pairs of alternate positional arguments. So that would not fall under the question of "given a flag with an argument named `NAME` in the usage, return the `NAME`."

**Comment by con-f-use** [comment]
I see. Getting the "type" of positional arguments, however, would be the logical extention of the SO request. It might also be nice for my beloved autocompletion. If it is known something has a `FILE` or `HOSTNAME` or `USER` "type" argument, one can uses the appropriate pre-defined completion functions that come with bash.

**Comment by BurntSushi** [comment]
In this case, the "type" of the positional argument is just its name, which is available in standard Docopt. The key is that the name of a *flag's* argument is dropped on the floor (because it has no semantic value other than  its existence).

**Comment by con-f-use** [comment]
You're right, I didn't think. Sorry. For sake of staying on topic consider deleting this and your previous post.

**Comment by jkgeyti** [comment]
Question is whether it, in con-f-use's example, should be possible to determine that the type of the positional arguments as `[ ("<cartesian_x>","<cartesian_y>"), ("<polar_angle>", "<polar_radius>") ]`, i.e. a list of everything it could possibly be resolved to. I'm not sure whether that in itself adds much value, but being able to resolve positional arguments definitely does.

**Comment by BurntSushi** [comment]
Nah, this is good discussion! I had to go back and re-read the SO post myself, so having clarification in more detail is good. Thank you! :-)
